fvm/meter/interaction_meter.go

```diff
@@ -51,34 +57,55 @@ func (m *InteractionMeter) MeterStorageRead(
 ) error {
 
 	// all reads are on a View which only read from storage at the first read of a given key
-	if _, ok := m.storageUpdateSizeMap[storageKey]; !ok {
+	if _, ok := m.reads[storageKey]; !ok {
 		readByteSize := getStorageKeyValueSize(storageKey, value)
 		m.totalStorageBytesRead += readByteSize
-		m.storageUpdateSizeMap[storageKey] = readByteSize
+		m.reads[storageKey] = readByteSize
 	}
 
 	return m.checkStorageInteractionLimit(enforceLimit)
 }
 
 // MeterStorageWrite captures storage written bytes count and returns an error
-// if it goes beyond the total interaction limit and limit is enforced
+// if it goes beyond the total interaction limit and limit is enforced.
+// If a key is written multiple times, only the last write is counted.
+// If a key is written before it has been read, next time it will be read it will be from the view,
+// not from storage, so count it as read 0.
 func (m *InteractionMeter) MeterStorageWrite(
 	storageKey flow.RegisterID,
 	value flow.RegisterValue,
 	enforceLimit bool,
 ) error {
-	// all writes are on a View which only writes the latest updated value to storage at commit
-	if old, ok := m.storageUpdateSizeMap[storageKey]; ok {
-		m.totalStorageBytesWritten -= old
-	}
-
 	updateSize := getStorageKeyValueSize(storageKey, value)
-	m.totalStorageBytesWritten += updateSize
-	m.storageUpdateSizeMap[storageKey] = updateSize
+	m.replaceWrite(storageKey, updateSize)
+
+	if _, ok := m.reads[storageKey]; !ok {
+		// write without read, count as read 0 because next time you read it the written value will be returned
+		// so no interaction with storage will happen.
+		m.reads[storageKey] = 0
+	}
 
 	return m.checkStorageInteractionLimit(enforceLimit)
 }
 
+// replaceWrite replaces the write size of a given key with the new size, because
+// only the last write of a given key is counted towards the total interaction limit.
+// These are the only write usages of `m.totalStorageBytesWritten` and `m.writes`,
+// which means that `m.totalStorageBytesWritten` can never become negative.
+// oldSize is always <= m.totalStorageBytesWritten.
+func (m *InteractionMeter) replaceWrite(
+	k flow.RegisterID,
+	newSize uint64,
+) {
+	// remove old write
+	oldSize := m.writes[k]
+	m.totalStorageBytesWritten -= oldSize
```

Comment:
can we add a sanity check that after this operation, `m.totalStorageBytesWritten` is not greater than before.

Reply:
I added a log instead of an error, as the error would be difficult to propagate out (the `Merge` function intentionally doesn't return an error).

I also set the `totalStorageBytesWritten` to 0 in that case. If this does happen (I still don't see how), worst that will happen is that the interaction is measured incorrectly.